index/scorch/persister.go

```diff
@@ -871,30 +888,124 @@ func (s *Scorch) removeOldData() {
 // rollback'ability.
 var NumSnapshotsToKeep = 1
 
+// RollbackSamplingInterval controls how far back we are looking
+// in the history to get the rollback points.
+// For example, a value of 10 minutes ensures that the
+// protected snapshots (NumSnapshotsToKeep = 3) are:
+//    the very latest snapshot(ie the current one),
+//    the snapshot that was persisted 10 minutes before the current one,
+//	  the snapshot that was persisted 20 minutes before the current one
+//
+// By default however, the timeseries way of protecting snapshots is
+// disabled, and we protect the latest three contiguous snapshots
+var RollbackSamplingInterval = 0 * time.Minute
+
+// Controls what portion of the earlier rollback points to retain during
+// a infrequent/sparse mutation scenario
+var RollbackRetentionFactor = float64(0.5)
+
+func getTimeSeriesSnapshots(maxDataPoints int, interval time.Duration,
+	snapshots []*snapshotMetaData) (int, map[uint64]time.Time) {
+	if interval == 0 {
+		return len(snapshots), map[uint64]time.Time{}
+	}
+	// the map containing the time series snapshots, i.e the timeseries of snapshots
+	// each of which is separated by rollbackSamplingInterval
+	rv := make(map[uint64]time.Time)
+	// the last point in the "time series", i.e. the timeseries of snapshots
+	// each of which is separated by rollbackSamplingInterval
+	ptr := len(snapshots) - 1
+	rv[snapshots[ptr].epoch] = snapshots[ptr].timeStamp
+	numSnapshotsProtected := 1
+
+	// traverse the list in reverse order, older timestamps to newer ones.
+	for i := ptr - 1; i >= 0; i-- {
+		// If we find a timeStamp which is the next datapoint in our
+		// timeseries of snapshots, and newer by RollbackSamplingInterval duration
+		// (comparison in terms of minutes), which is the interval of our time
+		// series. In this case, add the epoch rv
+		if int(snapshots[i].timeStamp.Sub(snapshots[ptr].timeStamp).Minutes()) >
+			int(interval.Minutes()) {
```

Comment:
Why cast both the Minutes() above into int - we can simply leave them in the float64 format as it is just a greater than comparison.

Reply:
Done